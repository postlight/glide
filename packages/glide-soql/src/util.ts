export const IS_EMPTY = Symbol("glide.isEmpty");

export interface Fragment {
  [IS_EMPTY](): boolean;
  toString(): string;
}

export function notEmpty(input: Fragment): boolean {
  return !isEmpty(input);
}

export function isEmpty(input: Fragment): boolean {
  return input[IS_EMPTY]();
}
