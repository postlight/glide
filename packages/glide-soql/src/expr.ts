import * as ops from "./ops";
import { IS_EMPTY, Fragment } from "./util";

const SINGLE_QUOTE = /'/gm;

export type Expr = Bool | Cmp;
export type Field = string | Fn;
export type Scalar = bigint | boolean | null | number | string;

export class Bool implements Fragment {
  private readonly value: Expr[];

  private constructor(private readonly operator: ops.Bool, value: Expr[]) {
    this.value = value.reduce<Expr[]>((acc, item) => {
      if (item instanceof Cmp || operator !== item.operator) {
        acc.push(item);
        return acc;
      }

      return acc.concat(item.value);
    }, []);
  }

  static and(...value: Expr[]): Bool {
    return new Bool(ops.Bool.And, value);
  }

  static or(...value: Expr[]): Bool {
    return new Bool(ops.Bool.Or, value);
  }

  [IS_EMPTY](): boolean {
    return this.value.length === 0;
  }

  toString(): string {
    return this.value
      .map(expr => (expr instanceof Bool ? `(${expr})` : `${expr}`))
      .join(` ${this.operator} `);
  }
}

export class Cmp implements Fragment {
  private constructor(
    private readonly field: Field,
    private readonly operator: ops.Cmp,
    private readonly expression: Scalar | Scalar[],
  ) {}

  static eq(field: Field, expression: Scalar): Cmp {
    return new Cmp(field, ops.Cmp.Eq, expression);
  }

  static gt(field: Field, expression: Scalar): Cmp {
    return new Cmp(field, ops.Cmp.Gt, expression);
  }

  static gte(field: Field, expression: Scalar): Cmp {
    return new Cmp(field, ops.Cmp.Gte, expression);
  }

  static in(field: Field, expression: Scalar[]): Cmp {
    return new Cmp(field, ops.Cmp.In, expression);
  }

  static like(field: Field, expression: string): Cmp {
    return new Cmp(field, ops.Cmp.Like, expression);
  }

  static lt(field: Field, expression: Scalar): Cmp {
    return new Cmp(field, ops.Cmp.Lt, expression);
  }

  static lte(field: Field, expression: Scalar): Cmp {
    return new Cmp(field, ops.Cmp.Lte, expression);
  }

  static ne(field: Field, expression: Scalar): Cmp {
    return new Cmp(field, ops.Cmp.Ne, expression);
  }

  static notIn(field: Field, expression: Scalar[]): Cmp {
    return new Cmp(field, ops.Cmp.NotIn, expression);
  }

  [IS_EMPTY](): boolean {
    return false;
  }

  toString(): string {
    let expression: string;

    if (Array.isArray(this.expression)) {
      expression = `(${this.expression.map(stringifyScalar).join(", ")})`;
    } else {
      expression = stringifyScalar(this.expression);
    }

    return `${this.field} ${this.operator} ${expression}`;
  }
}

export class Fn implements Fragment {
  private constructor(private readonly ident: string, private readonly field?: string) {}

  static avg(field: string): Fn {
    return new Fn("AVG", field);
  }

  static calendarMonth(field: string): Fn {
    return new Fn("CALENDAR_MONTH", field);
  }

  static calendarQuarter(field: string): Fn {
    return new Fn("CALENDAR_QUARTER", field);
  }

  static calendarYear(field: string): Fn {
    return new Fn("CALENDAR_YEAR", field);
  }

  static convertTimezone(field: string): Fn {
    return new Fn("CONVERT_TIMEZONE", field);
  }

  static count(field?: string): Fn {
    return new Fn("COUNT", field);
  }

  static countDistinct(field?: string): Fn {
    return new Fn("COUNT_DISTINCT", field);
  }

  static dayInMonth(field: string): Fn {
    return new Fn("DAY_IN_MONTH", field);
  }

  static dayInWeek(field: string): Fn {
    return new Fn("DAY_IN_WEEK", field);
  }

  static dayInYear(field: string): Fn {
    return new Fn("DAY_IN_YEAR", field);
  }

  static dayOnly(field: string): Fn {
    return new Fn("DAY_ONLY", field);
  }

  static fiscalMonth(field: string): Fn {
    return new Fn("FISCAL_MONTH", field);
  }

  static fiscalQuarter(field: string): Fn {
    return new Fn("FISCAL_QUARTER", field);
  }

  static fiscalYear(field: string): Fn {
    return new Fn("FISCAL_YEAR", field);
  }

  static hourInDay(field: string): Fn {
    return new Fn("HOUR_IN_DAY", field);
  }

  static max(field: string): Fn {
    return new Fn("MAX", field);
  }

  static min(field: string): Fn {
    return new Fn("MIN", field);
  }

  static sum(field: string): Fn {
    return new Fn("SUM", field);
  }

  static weekInMonth(field: string): Fn {
    return new Fn("WEEK_IN_MONTH", field);
  }

  static weekInYear(field: string): Fn {
    return new Fn("WEEK_IN_YEAR", field);
  }

  [IS_EMPTY](): boolean {
    return false;
  }

  toString(): string {
    return `${this.ident}(${this.field ? this.field : ""})`;
  }
}

function stringifyScalar(input: Scalar): string {
  switch (typeof input) {
    case "boolean":
      return `${input}`.toUpperCase();
    case "string":
      return `'${input.replace(SINGLE_QUOTE, "\\'")}'`;
    default:
      return input == null ? "null" : String(input);
  }
}
