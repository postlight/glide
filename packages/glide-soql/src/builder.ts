import { Direction, GroupBy, Limit, Offset, OrderBy, Where } from "./clause";
import { Bool, Cmp, Field } from "./expr";
import { IS_EMPTY, Fragment, notEmpty } from "./util";

export interface Parts {
  groupBy: GroupBy;
  limit: Limit;
  offset: Offset;
  orderBy: OrderBy;
  select: Select;
  where: Where;
}

export class Builder {
  private readonly parts: Parts;

  constructor(select: Select) {
    this.parts = {
      select,
      groupBy: new GroupBy(),
      limit: new Limit(),
      offset: new Offset(),
      orderBy: new OrderBy(),
      where: new Where(),
    };
  }

  static isBuilder(value: unknown): value is Builder {
    return value instanceof Builder;
  }

  groupBy(field: Field | Field[], having?: Bool | Cmp): Builder {
    const fields = Array.isArray(field) ? field : [field];

    this.parts.groupBy = new GroupBy(fields, having);
    return this;
  }

  limit(size: bigint | number): Builder {
    this.parts.limit = new Limit(size);
    return this;
  }

  offset(size: bigint | number): Builder {
    this.parts.offset = new Offset(size);
    return this;
  }

  orderBy(field: Field | Field[], direction?: Direction): Builder {
    const fields = Array.isArray(field) ? field : [field];

    this.parts.orderBy = new OrderBy(fields, direction);
    return this;
  }

  select(...fields: (Field | Select)[]): Builder {
    this.parts.select.select(...fields);
    return this;
  }

  toString(): string {
    const parts = this.parts;
    let query = `${parts.select}`;

    if (query === "") {
      return query;
    }

    if (notEmpty(parts.where)) {
      query += ` ${parts.where}`;
    }

    if (notEmpty(parts.groupBy)) {
      query += ` ${parts.groupBy}`;
    }

    if (notEmpty(parts.orderBy)) {
      query += ` ${parts.orderBy}`;
    }

    if (notEmpty(parts.limit)) {
      query += ` ${parts.limit}`;
    }

    if (notEmpty(parts.offset)) {
      query += ` ${parts.offset}`;
    }

    return query;
  }

  where(condition: Bool | Cmp): Builder {
    this.parts.where = new Where(condition);
    return this;
  }
}

export class Select implements Fragment {
  private readonly fields = new Set<Field | Select>();

  private constructor(private readonly object?: string) {}

  static from(object: string): Select {
    return new Select(object);
  }

  static isSelect(value: unknown): value is Select {
    return value instanceof Select;
  }

  [IS_EMPTY](): boolean {
    return this.fields.size === 0;
  }

  concat({ fields }: Select): Select {
    return this.select(
      ...Array.from(fields, field => {
        if (field instanceof Select) {
          const { fields, object } = field;
          const subselect = new Select(`${this.object}.${object}`);

          return subselect.select(...fields);
        }

        return `${this.object}.${field}`;
      }),
    );
  }

  select(...fields: (Field | Select)[]): Select {
    fields.forEach(field => this.fields.add(field));
    return this;
  }

  toString(): string {
    const fields = Array.from(this.fields)
      .filter(field => (typeof field === "string" ? Boolean(field) : notEmpty(field)))
      .map(field => (field instanceof Select ? `(${field})` : `${field}`));

    return fields.length > 0 ? `SELECT ${fields.join(", ")} FROM ${this.object}` : "";
  }
}
