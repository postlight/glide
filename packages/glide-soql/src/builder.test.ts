import { Builder, Select } from "./builder";
import { Bool, Cmp } from "./expr";
import { isEmpty } from "./util";

describe("Builder", () => {
  let subject: Builder;

  beforeEach(() => {
    subject = new Builder(Select.from("User"));
  });

  test("#groupBy()", () => {
    subject.groupBy("LastName");
    expect(subject).toMatchSnapshot();

    subject.groupBy(["LastName", "FirstName"]);
    expect(subject).toMatchSnapshot();
  });

  test("#limit()", () => {
    subject.limit(10);
    expect(subject).toMatchSnapshot();
  });

  test("#offset()", () => {
    subject.offset(10);
    expect(subject).toMatchSnapshot();
  });

  test("#orderBy()", () => {
    subject.orderBy("FirstName");
    expect(subject).toMatchSnapshot();

    subject.orderBy(["FirstName", "LastName"]);
    expect(subject).toMatchSnapshot();
  });

  test("#select()", () => {
    subject.select("FirstName", "LastName");
    expect(subject).toMatchSnapshot();
  });

  test("#toString()", () => {
    const empty = new Builder(Select.from("User"));
    const notEmpty = new Builder(Select.from("User"))
      .select("Id")
      .where(
        Bool.and(
          Cmp.eq("Company", "Postlight"),
          Bool.or(
            Cmp.eq("EmailAddress", "hello@postlight.com"),
            Cmp.eq("PhoneNumber", "+16466948530"),
          ),
        ),
      )
      .groupBy("LastName")
      .orderBy("LastName")
      .limit(10)
      .offset(10);

    expect(`${empty}`).toMatchSnapshot();
    expect(`${notEmpty}`).toMatchSnapshot();
  });

  test("#where()", () => {
    subject.where(Cmp.eq("Id", null));
    expect(subject).toMatchSnapshot();
  });
});

describe("Select", () => {
  let subject: Select;

  beforeEach(() => {
    subject = Select.from("User");
  });

  test("#@@isEmpty()", () => {
    expect(isEmpty(subject)).toBe(true);
    expect(isEmpty(subject.select("Id"))).toBe(false);
  });

  test("#concat()", () => {
    const target = Select.from("Contact").select("Contact.FirstName", "Contact.LastName");
    expect(subject.concat(target)).toMatchSnapshot();
  });

  test("#select()", () => {
    subject.select("FirstName", "LastName");
    expect(subject).toMatchSnapshot();
  });

  test("#toString()", () => {
    const simple = Select.from("User").select("FirstName", "LastName");
    const complex = Select.from("User").select("Id", Select.from("Contact").select("Id"));

    expect(`${simple}`).toMatchSnapshot();
    expect(`${complex}`).toMatchSnapshot();
  });
});
