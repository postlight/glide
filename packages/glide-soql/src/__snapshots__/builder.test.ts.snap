// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Builder #groupBy() 1`] = `
Builder {
  "parts": Object {
    "groupBy": GroupBy {
      "fields": Array [
        "LastName",
      ],
      "having": undefined,
    },
    "limit": Limit {
      "size": undefined,
    },
    "offset": Offset {
      "size": undefined,
    },
    "orderBy": OrderBy {
      "direction": "ASC",
      "fields": Array [],
    },
    "select": Select {
      "fields": Set {},
      "object": "User",
    },
    "where": Where {
      "condition": undefined,
    },
  },
}
`;

exports[`Builder #groupBy() 2`] = `
Builder {
  "parts": Object {
    "groupBy": GroupBy {
      "fields": Array [
        "LastName",
        "FirstName",
      ],
      "having": undefined,
    },
    "limit": Limit {
      "size": undefined,
    },
    "offset": Offset {
      "size": undefined,
    },
    "orderBy": OrderBy {
      "direction": "ASC",
      "fields": Array [],
    },
    "select": Select {
      "fields": Set {},
      "object": "User",
    },
    "where": Where {
      "condition": undefined,
    },
  },
}
`;

exports[`Builder #limit() 1`] = `
Builder {
  "parts": Object {
    "groupBy": GroupBy {
      "fields": Array [],
      "having": undefined,
    },
    "limit": Limit {
      "size": 10,
    },
    "offset": Offset {
      "size": undefined,
    },
    "orderBy": OrderBy {
      "direction": "ASC",
      "fields": Array [],
    },
    "select": Select {
      "fields": Set {},
      "object": "User",
    },
    "where": Where {
      "condition": undefined,
    },
  },
}
`;

exports[`Builder #offset() 1`] = `
Builder {
  "parts": Object {
    "groupBy": GroupBy {
      "fields": Array [],
      "having": undefined,
    },
    "limit": Limit {
      "size": undefined,
    },
    "offset": Offset {
      "size": 10,
    },
    "orderBy": OrderBy {
      "direction": "ASC",
      "fields": Array [],
    },
    "select": Select {
      "fields": Set {},
      "object": "User",
    },
    "where": Where {
      "condition": undefined,
    },
  },
}
`;

exports[`Builder #orderBy() 1`] = `
Builder {
  "parts": Object {
    "groupBy": GroupBy {
      "fields": Array [],
      "having": undefined,
    },
    "limit": Limit {
      "size": undefined,
    },
    "offset": Offset {
      "size": undefined,
    },
    "orderBy": OrderBy {
      "direction": "ASC",
      "fields": Array [
        "FirstName",
      ],
    },
    "select": Select {
      "fields": Set {},
      "object": "User",
    },
    "where": Where {
      "condition": undefined,
    },
  },
}
`;

exports[`Builder #orderBy() 2`] = `
Builder {
  "parts": Object {
    "groupBy": GroupBy {
      "fields": Array [],
      "having": undefined,
    },
    "limit": Limit {
      "size": undefined,
    },
    "offset": Offset {
      "size": undefined,
    },
    "orderBy": OrderBy {
      "direction": "ASC",
      "fields": Array [
        "FirstName",
        "LastName",
      ],
    },
    "select": Select {
      "fields": Set {},
      "object": "User",
    },
    "where": Where {
      "condition": undefined,
    },
  },
}
`;

exports[`Builder #select() 1`] = `
Builder {
  "parts": Object {
    "groupBy": GroupBy {
      "fields": Array [],
      "having": undefined,
    },
    "limit": Limit {
      "size": undefined,
    },
    "offset": Offset {
      "size": undefined,
    },
    "orderBy": OrderBy {
      "direction": "ASC",
      "fields": Array [],
    },
    "select": Select {
      "fields": Set {
        "FirstName",
        "LastName",
      },
      "object": "User",
    },
    "where": Where {
      "condition": undefined,
    },
  },
}
`;

exports[`Builder #toString() 1`] = `""`;

exports[`Builder #toString() 2`] = `"SELECT Id FROM User WHERE Company = 'Postlight' AND (EmailAddress = 'hello@postlight.com' OR PhoneNumber = '+16466948530') GROUP BY LastName ORDER BY LastName ASC LIMIT 10 OFFSET 10"`;

exports[`Builder #where() 1`] = `
Builder {
  "parts": Object {
    "groupBy": GroupBy {
      "fields": Array [],
      "having": undefined,
    },
    "limit": Limit {
      "size": undefined,
    },
    "offset": Offset {
      "size": undefined,
    },
    "orderBy": OrderBy {
      "direction": "ASC",
      "fields": Array [],
    },
    "select": Select {
      "fields": Set {},
      "object": "User",
    },
    "where": Where {
      "condition": Cmp {
        "expression": null,
        "field": "Id",
        "operator": "=",
      },
    },
  },
}
`;

exports[`Select #concat() 1`] = `
Select {
  "fields": Set {
    "User.Contact.FirstName",
    "User.Contact.LastName",
  },
  "object": "User",
}
`;

exports[`Select #select() 1`] = `
Select {
  "fields": Set {
    "FirstName",
    "LastName",
  },
  "object": "User",
}
`;

exports[`Select #toString() 1`] = `"SELECT FirstName, LastName FROM User"`;

exports[`Select #toString() 2`] = `"SELECT Id, (SELECT Id FROM Contact) FROM User"`;
